### ISSUE TITLE
Add copy-editing of metagenotype annotations

### ISSUE THREAD

**Issue by jseager7** [pombase/canto#1760]
(Requested by @CuzickA)

Currently, there is no way to change the target metagenotype of a metagenotype annotation (_pathogen&ndash;host interaction phenotype_). Instead, a new annotation has to be created each time, which is a problem for experiments where many interactions display the same phenotypes.

This is how the annotation editing popup looks now:

<img width="573" alt="mg-ant-no-selector" src="https://user-images.githubusercontent.com/37659591/52274267-0af73980-2944-11e9-889b-29b059ca6c82.PNG">

Ideally, there should be a selector listing the metagenotypes added to the session. The organism name may have to be displayed in this list as well, in the case that different organisms have the same gene names. We can either re-use the existing text display for metagenotypes (as seen in the popup title in the image above), or we can use something more compact, for example:

> TRI5&Delta; (F. graminearum, PH-1) &mdash;  OPR3&Delta; (S. lycopersicum, Craigella)

Which would look like this in the selector:

<img width="577" alt="mg-ant-add-selector" src="https://user-images.githubusercontent.com/37659591/52275514-ad64ec00-2947-11e9-92b9-25a59d370509.PNG">

- - -

Note that the metagenotype name may have to account for alleles with more complicated formatting &ndash; using standard parentheses to delimit the organism name and strain might be confusing when those parentheses are in the allele names as well, for example with nonsense mutations:

> unnamed(aa123->stop)

We could solve this by using parentheses that are never used in allele names, for example square brackets `[]` or curly brackets `{}`.

**Comment by jseager7** [comment]
### Relevant code:

- Template `annotation_edit.html`
- Controller `annotationEditDialogCtrl`
- Function `startEditing` (opens modal linked to template and controller)
- Directive `featureChooser`
  - presumably this needs the `featureType` binding to have a value of `"metagenotype"`, but the controller doesn't support this option yet; the `Metagenotype` service needs to be injected and logic needs to be added to `function featureChooserControlHelper`.

**Comment by ValWood** [comment]
Yes this is strange In normal Canto you *can* edit the genotype in this pop up with cop/edit.....

**Comment by kimrutherford** [comment]
> Yes this is strange In normal Canto you can edit the genotype in this pop up with cop/edit.....

Changing the genotype still works.  It's changing the meta-genotype of an annotation that needs implementing.

**Comment by kimrutherford** [comment]
I've changed the Perl code that returns metagenotype details so that a minimal display_name is returned for each metagenotype.  Currently it's just pathogen and host display names separated by "/".

**Comment by kimrutherford** [comment]
> minimal display_name is returned

I've also added a `feature_id` that's needed by the feature-chooser.

**Comment by kimrutherford** [comment]
While trying to understand #1768 I think I've implemented some of this issue.  I've made pull request: https://github.com/pombase/canto/pull/1769

**Comment by jseager7** [comment]
OK, the metagenotypes display in the feature chooser now:

![1760-metagenotype-dropdown](https://user-images.githubusercontent.com/37659591/52716159-137bef80-2f96-11e9-98a2-4454f4d30f69.PNG)

Unfortunately, **the chooser should really display the organism and strain names** (as suggested above), since the most common use case for copying pathogen&ndash;host annotations is copying them onto different strains of the same organism, and those metagenotypes can't be distinguished at the moment. I'd like it to look something like this:

![1760-metagenotype-dropdown_fmt](https://user-images.githubusercontent.com/37659591/52717410-fa287280-2f98-11e9-9fb9-1de6c870c9ea.PNG)

@kimrutherford I'm not sure how easy it is to change the metagenotype feature formatting, since the formatting seems to be handled by `featureChooserFilter`, and I'm not sure if adding conditional filtering there for metagenotypes will have unexpected side effects.

**Comment by jseager7** [comment]
After further investigation, it looks like I could add some conditional formatting to `featureChooserFilter` based on whether `feature.metagenotype_id` is present:

```js
canto.filter('featureChooserFilter', function () {
  return function (feature) {
    var ret = ""
    if (feature.metagenotype_id === undefined) {
      ret = feature.display_name;
    } else {
      /* some other formatting rule */
    }
    if (feature.background) {
      ret += "  (bkg: " + feature.background.substr(0, 15);
      if (feature.background.length > 15) {
        ret += " ...";
      }
      ret += ")";
    }
    return ret;
  };
});
```

But maybe we should just change the display name of metagenotypes to include the organism and strain instead.

**Comment by kimrutherford** [comment]
Hi James.

The text displayed in the chooser is the "display_name" field of the meta-genotype data returned from the server.  It's quick to change the display_name.  I'm a bit concerned that the chooser will become too wide but I don't know how to avoid that.

**Comment by kimrutherford** [comment]
I've added the organism name and strain to the "display name" of the metagenotypes.  In Firefox the annotaion edit dialog looks like this with some long-ish names.  It handles things better than I thought it would.

![metagenotype-edit-long-names-1](https://user-images.githubusercontent.com/90474/53378295-fbf72a80-39c9-11e9-8918-3a4c1d4ff5b6.png)

**Comment by jseager7** [comment]
Those are some very long genotype names 😮

Good to see that even in the worse case it doesn't overflow your screen, but it could possibly overflow on smaller screens like laptops. I've got some ideas for how to save space, presuming we can add some conditional formatting for the feature chooser. Since this issue was only to do with adding the functionality, I'll open a new issue for improving the display. Closing.
